pkg/controller/upload-controller.go

```diff
@@ -490,37 +324,320 @@ func (c *UploadController) getOrCreateUploadService(pvc *v1.PersistentVolumeClai
+	pod := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: pvc.Namespace,
+			Annotations: map[string]string{
+				annCreatedByUpload: "yes",
+			},
+			Labels: map[string]string{
+				common.CDILabelKey:              common.CDILabelValue,
+				common.CDIComponentLabel:        common.UploadServerCDILabel,
+				common.UploadServerServiceLabel: name,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				MakePVCOwnerReference(pvc),
+			},
+		},
+		Spec: v1.PodSpec{
+			SecurityContext: &v1.PodSecurityContext{
+				RunAsUser: &[]int64{0}[0],
+			},
+			Containers: []v1.Container{
+				{
+					Name:            common.UploadServerPodname,
+					Image:           r.Image,
+					ImagePullPolicy: v1.PullPolicy(r.PullPolicy),
+					Env: []v1.EnvVar{
+						{
+							Name:  "TLS_KEY",
+							Value: string(serverKey),
+						},
+						{
+							Name:  "TLS_CERT",
+							Value: string(serverCert),
+						},
+						{
+							Name:  "CLIENT_CERT",
+							Value: string(clientCA),
+						},
+						{
+							Name:  common.UploadImageSize,
+							Value: requestImageSize,
+						},
+						{
+							Name:  "CLIENT_NAME",
+							Value: clientName,
+						},
+					},
+					Args: []string{"-v=" + r.Verbose},
+					ReadinessProbe: &v1.Probe{
+						Handler: v1.Handler{
+							HTTPGet: &v1.HTTPGetAction{
+								Path: "/healthz",
+								Port: intstr.IntOrString{
+									Type:   intstr.Int,
+									IntVal: 8080,
+								},
+							},
+						},
+						InitialDelaySeconds: 2,
+						PeriodSeconds:       5,
+					},
+				},
+			},
+			RestartPolicy: v1.RestartPolicyOnFailure,
+			Volumes: []v1.Volume{
+				{
+					Name: DataVolName,
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc.Name,
+							ReadOnly:  false,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if resourceRequirements != nil {
+		pod.Spec.Containers[0].Resources = *resourceRequirements
+	}
+
+	if getVolumeMode(pvc) == v1.PersistentVolumeBlock {
+		pod.Spec.Containers[0].VolumeDevices = addVolumeDevicesForUpload()
+		pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, v1.EnvVar{
+			Name:  "DESTINATION",
+			Value: common.WriteBlockPath,
+		})
+	} else {
+		pod.Spec.Containers[0].VolumeMounts = addVolumeMountsForUpload()
+	}
+
+	if scratchName != "" {
+		pod.Spec.Volumes = append(pod.Spec.Volumes, v1.Volume{
+			Name: ScratchVolName,
+			VolumeSource: v1.VolumeSource{
+				PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+					ClaimName: scratchName,
+					ReadOnly:  false,
+				},
+			},
+		})
+
+		pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts, v1.VolumeMount{
+			Name:      ScratchVolName,
+			MountPath: common.ScratchDataDir,
+		})
+	}
+
+	return pod
+}
+
+func addVolumeDevicesForUpload() []v1.VolumeDevice {
```

Comment:
this is only called once, maybe just put it inline?

Reply:
Sure.